src/include/buffer.h

```diff
@@ -585,6 +585,20 @@ namespace buffer CEPH_BUFFER_API {
 	  out_of_band_offset += p.length();
 	}
       }
+      void append(bufferptr&& p) {
+	if (!p.length()) {
+	  return;
+	}
+	if (deep) {
+	  append(p.c_str(), p.length());
+          p = bufferptr{}; // drop ref
+	} else {
+	  flush_and_continue();
+	  out_of_band_offset += p.length();
+	  pbl->append(std::move(p));
```

Comment:
What about using `buffer::list::push_back(ptr&&)` instead?  It's defined in `buffer.h` so compilers would have a chance to inline the code.